File: multinode/storage/storage.go
// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

package storage

import (
	"sort"
	"time"
)

// Usage holds storage usage stamps and summary for a particular period.
type Usage struct {
	Stamps       []UsageStamp `json:"stamps"`
	Summary      float64      `json:"summary"`
	SummaryBytes float64      `json:"summaryBytes"`
}

// UsageStamp holds data at rest total for an interval beginning at interval start.
type UsageStamp struct {
	AtRestTotal      float64   `json:"atRestTotal"`
	AtRestTotalBytes float64   `json:"atRestTotalBytes"`
	IntervalStart    time.Time `json:"intervalStart"`
}

// UsageStampDailyCache caches storage usage stamps by interval date.
type UsageStampDailyCache map[time.Time]UsageStamp

// Add adds usage stamp to cache aggregating at rest data by date.
func (cache *UsageStampDailyCache) Add(stamp UsageStamp) {
	year, month, day := stamp.IntervalStart.UTC().Date()
	intervalStart := time.Date(year, month, day, 0, 0, 0, 0, time.UTC)

	cached := *cache

	cacheStamp, ok := cached[intervalStart]
	if ok {
		cached[intervalStart] = UsageStamp{
			AtRestTotal:      cacheStamp.AtRestTotal + stamp.AtRestTotal,
			AtRestTotalBytes: cacheStamp.AtRestTotalBytes + stamp.AtRestTotalBytes,
			IntervalStart:    intervalStart,
		}
	} else {
		cached[intervalStart] = UsageStamp{
			AtRestTotal:      stamp.AtRestTotal,
			AtRestTotalBytes: stamp.AtRestTotalBytes,
			IntervalStart:    intervalStart,
		}
	}

	*cache = cached
}

// Sorted returns usage stamp slice sorted by interval start.
func (cache *UsageStampDailyCache) Sorted() []UsageStamp {
	var usage []UsageStamp

	for _, stamp := range *cache {
		usage = append(usage, stamp)
	}
	sort.Slice(usage, func(i, j int) bool {
		return usage[i].IntervalStart.Before(usage[j].IntervalStart)
	})

	return usage
}

// DiskSpace stores all info about storagenode disk space usage.
type DiskSpace struct {
	Allocated       int64 `json:"allocated"`
	Used            int64 `json:"used"`
	UsedPieces      int64 `json:"usedPieces"`
	UsedReclaimable int64 `json:"usedReclaimable"`
	UsedTrash       int64 `json:"usedTrash"`
	// Free is the actual amount of free space on the whole disk, not just allocated disk space, in bytes.
	Free int64 `json:"free"`
	// Available is the amount of free space on the allocated disk space, in bytes.
	Available int64 `json:"available"`
	Overused  int64 `json:"overused"`
}

// Add combines disk space with another one.
func (diskSpace *DiskSpace) Add(space DiskSpace) {
	diskSpace.Allocated += space.Allocated
	diskSpace.Used += space.Used
	diskSpace.UsedTrash += space.UsedTrash
	diskSpace.UsedReclaimable += space.UsedReclaimable
	diskSpace.UsedPieces += space.UsedPieces
	diskSpace.Free += space.Free
	diskSpace.Available += space.Available
	diskSpace.Overused += space.Overused
}

File: web/multinode/src/api/storage.ts
// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import { APIClient } from '@/api/index';
import { DiskSpace, DiskSpaceUsage, Stamp } from '@/storage';

/**
 * Client for storage controller of MND api.
 */
export class StorageClient extends APIClient {
    private readonly ROOT_PATH: string = '/api/v0/storage';

    /**
     * Returns storage usage information for selected node and satellite if any.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the auth cookie is missing or invalid.
     *
     * @throws {@link InternalError}
     * Thrown if something goes wrong on server side.
     */
    public async usage(satelliteId: string | null, nodeId: string | null): Promise<DiskSpaceUsage> {
        let path = `${this.ROOT_PATH}`;

        if (satelliteId) {
            path += `/satellites/${satelliteId}`;
        }

        path += '/usage';

        if (nodeId) {
            path += `/${nodeId}`;
        }

        const now = new Date();
        const year = now.getUTCFullYear();
        const month = now.getUTCMonth() + 1;
        const period = `${year}-${month > 9 ? month : `0${month}`}`;

        path += `?period=${period}`;

        const response = await this.http.get(path);

        if (!response.ok) {
            await this.handleError(response);
        }

        const data = await response.json();
        const usage = data.stamps || [];

        return new DiskSpaceUsage(
            usage.map(stamp => new Stamp(stamp.atRestTotal, stamp.atRestTotalBytes, new Date(stamp.intervalStart))),
            data.summary,
            data.summaryBytes,
        );
    }

    /**
     * Returns disk space information for selected node if selected.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the auth cookie is missing or invalid.
     *
     * @throws {@link InternalError}
     * Thrown if something goes wrong on server side.
     */
    public async diskSpace(nodeId: string | null): Promise<DiskSpace> {
        let path = `${this.ROOT_PATH}/disk-space`;

        if (nodeId) {
            path += `/${nodeId}`;
        }

        const response = await this.http.get(path);

        if (!response.ok) {
            await this.handleError(response);
        }

        const diskSpace = await response.json();

        return new DiskSpace(
            diskSpace.allocated,
            diskSpace.used,
            diskSpace.usedPieces,
            diskSpace.usedReclaimable,
            diskSpace.usedTrash,
            diskSpace.free,
            diskSpace.available,
            diskSpace.overused,
        );
    }
}
